### DISCUSSION TITLE
key wrap decrypt failed

### DISCUSSION THREAD

**Discussion by laogaolao** [openssl/openssl#22861] [Q&A] [answered]
hi,
We are using the openssl 3.0 with fips provider. we met an issue when using openssl enc do the key wrap.
the command is:
**root@System:/tmp# openssl enc -id-aes256-wrap-pad -pbkdf2 -in upload_DB.zip -out test.zip -e -k aabbccdd -p -iv 1234
hex string is too short, padding with zero bytes to length
salt=946826E826244ED4
key=1AACC74C6A32CD810E4E4DF923B99985549A33B1CCCE5BFB805AD17FFC79CCBE
iv =12340000**
This execution is good without any error, but I try to decrypt using command:

I tried to decrypt this file, then got error:
**root@System:/tmp# openssl enc -id-aes256-wrap-pad -pbkdf2 -in test.zip -out test1.zip -d -k aabbccdd -p -iv 1234
hex string is too short, padding with zero bytes to length
salt=946826E826244ED4
key=1AACC74C6A32CD810E4E4DF923B99985549A33B1CCCE5BFB805AD17FFC79CCBE
iv =12340000
error writing output file
B7E58020:error:1C800066:Provider routines:aes_wrap_cipher_internal:cipher operation failed:../openssl-3.0.12/providers/implementations/ciphers/cipher_aes_wrp.c:192:
B7E58020:error:030000BD:digital envelope routines:EVP_DecryptUpdate:update error:../openssl-3.0.12/crypto/evp/evp_enc.c:837:**

Can you help check why we got this error? Thanks.

**Comment by t8m** [comment, marked as answer]
The wrap ciphers have only limited support in openssl enc. One limitation is that you cannot encrypt anything bigger than 4096 bytes.

May I ask why do you want to use a wrap cipher mode with enc? It is not supposed to be used to encrypt any data, its purpose is to encrypt raw keys - i.e., mostly random data.

**Reply by laogaolao** [reply]
thank you for the reply. Currently in our system we do the AES encryption with ECB mode for the password and key and these will saved in the DB. But this DB will be uploaded outside of our system, our SE said it does not meet the requirements of FIPS. So I think maybe I can use the key wrap to the whole DB.
I also tried same command with same file on the x86 ubuntu VM, it worked, but the openssl version is a little different with the versoin in the target system. In my VM, it's 3.0.13, in our target system, it's 3.0.12.

**Reply by laogaolao** [reply]
And the encryption is done but the decryption failed.

**Reply by t8m** [reply]
If you're encrypting a zip file, using AES-CBC might be a better idea.

**Reply by mattcaswell** [reply]
> So I think maybe I can use the key wrap to the whole DB.

Key wrap is for encrypting an individual key. It should not be used for a database (even if that database contains keys!).

**Reply by laogaolao** [reply]
Understood, thanks.

**Reply by suryanshsangwan** [reply]
ML-DSA-65 and ML-DSA-87 keys are larger than 4096 bytes. There should be a way to wrap them.

**Reply by slontis** [reply]
> ML-DSA-65 and ML-DSA-87 keys are larger than 4096 bytes. There should be a way to wrap them.

This is not good, I will look into it. One of the problems is that it uses a cipher BIO that applies this limit, but I can see a few other issues that need addressing also.
